install.sh

```diff
@@ -1,6 +1,12 @@
 #!/bin/bash
 set -u
 
+# Check if script is run non-interactively (e.g. CI)
+# If it is run non-interactively we should not prompt for passwords.
+if [[ ! -t 0 || ! -t 1 || $SHLVL == 1 ]] || ! tty -s; then
+  NONINTERACTIVE=1
+fi
```

Comment:
Can you try to remove this check and instead we adjust the `sudo` logic accordingly so it does not block?

Reply:
If you have any ideas?

Like I said before, there isn't really any way I can think off.

Also, it just boils down to, either you install it on your own computer in an interactive terminal and want to be prompted, or you install it non-interactively and don't want to be prompted. This is the clearest way to distinguish that. 

You could investigate the /etc/sudoers file and see what it informs you, but I am afraid there will non-interactive installs (that don't need to prompted) that have very similar /etc/sudoers files than interactive installs (that need to be prompted), so what better check than just check, "Have terminal" ? prompt : don't prompt.

So, what exactly do you don't like about the NONINTERACTIVE variable? We could perhaps change the if NONINTERACTIVE: return 0 block in have_sudo_access to args += "-n" (non-interactive), that might be a bit cleaner and so reusing the sudo logic, so it just makes the same checks, but just doesn't prompt? (I figured I could just do an early return, because it seems that on CI you just want the sudo install anyway.) We could even just bring all this logic to the have_sudo_access with if [[ ! -t 0 ]]; then args += "-n", but having a variable has the upside that you can set it yourself manually before running the install script. (If you are really sure you don't want to be prompted anyways, even if you are on a terminal.)

Reply:
> So, what exactly do you don't like about the NONINTERACTIVE variable?

That it's wasn't necessary before the rest of the changes in here and that I'm not convinced it will be comprehensive enough to work consistently.



> return 0 block in have_sudo_access to args += "-n" (non-interactive), that might be a bit cleaner and so reusing the sudo logic, so it just makes the same checks, but just doesn't prompt?

Yes, that makes sense to me.

Reply:
>That it's wasn't necessary before the rest of the changes in here and that I'm not convinced it will be comprehensive enough to work consistently.

The change that this PR primarily brought is more interactivity. So, it only makes sense that if we are making the script more interactive, more fine grained control on when or when not this interactivity must be activated.

----

I think the -t 0 is  already **very** strict and comprehensive if I understand its implementation correctly. It makes the check if the file descriptor (FD 0 is stdin) is connected to a **terminal**, in which the **terminal** used to be an actual hardware device talking to the TTY kernel module. Nowadays hardware **terminals** are rarely in use, and anyone uses a **virtual terminal**, xterm, urxvt, iTerm2 (but also tmux and sshd). But these **terminal emulators** should still **emulate** this hardware behavior to the kernel to make it be a terminal (and have terminal features). Technically it is thus possible for any piece of software to **emulate** being a **terminal** even though there was no human interacting with it, however it would in that case not make any sense for that piece of software to **emulate** being a **terminal**. So, that is why -t 0 is in my opinion a very comprehensive check to know if there is indeed a human there to see the prompt and act upon it.  

In fact the check -t 0 is already in use in the install script! `if [[ -t 0 && -z "${CI-}" ]]; then wait_for_user; fi`, and also in the uninstall script. `if [[ -t 0 && -z $opt_force && -z $opt_dry_run ]]; then read -rp "Are you sure you want to uninstall Homebrew? This will remove your installed packages! [y/N] ";  [[ $REPLY == [yY]* ]] || abort; fi`. 

----

Which brings me to the realization that there already exists a variable that can track this, `CI`. Didn't know that already. It is also not set automatically, so can only be set externally. I wonder if this is documented somewhere. 

So, we might be using CI instead of  NONINTERACTIVE (and than still set it automatically). However, if the variable CI is set it also unconditionally installs in `/home/linuxbrew/.linuxbrew` (because there is check that does that) without any way to install it somewhere else, which might not be desirable. So, we either have two variables, or remove that last check. (I have now kept both variables and added that if CI is set, NONINTERACTIVE is set as well.)

Reply:
> So, we might be using CI instead of NONINTERACTIVE (and than still set it automatically). However, if the variable CI is set it also unconditionally installs in `/home/linuxbrew/.linuxbrew` (because there is check that does that) without any way to install it somewhere else, which might not be desirable. So, we either have two variables, or remove that last check. (I have now kept both variables and added that if CI is set, NONINTERACTIVE is set as well.)

This approach makes sense to me given the explanation above, thanks 👍🏻.

The macOS CI is now failing, though.